### ISSUE TITLE
DicomClientSendingRequestsState holds onto the sent requests

### ISSUE THREAD

**Issue by jbergstr** [fo-dicom/fo-dicom#982]
### Expected behavior
The requests that have been sent are discarded and eventually garbage collected.

### Actual behavior
After a request is sent, `DicomClientSendingRequestsState` removes it from the queue and puts it to `ConcurrentBag<DicomRequest> _sentRequests`. This causes memory usage to blow up since the request objects cannot be garbage collected. `_sentRequests` seems to be only used for its `Count` property; the requests could certainly be counted with some other means, too.

### Steps to reproduce the behavior
Send a few hundred images in an association and the process will soon consume close to 1 GB of memory that is freed only after the association is released. This can also be seen using Visual Studio's Performance Profiler which shows an ever-increasing count of allocations eventually pointing to request objects that cannot be collected.

### fo-dicom version and OS/platform
fo-dicom 4.0.4

**Comment by gofal** [comment]
the sent Requests have to be stored in memory at least until the DicomResponse returned from the server. Because the `OnResponseReceived`-event passes both, the Request and the Response, as parameter. But you are right, after the OnResponseReceived-event has been handled the Request could be remove from memory.

**Comment by jbergstr** [comment]
Just in case anyone else runs into this problem, I've managed to work around it by introducing a "file scope" class which clears most of the data after the response has been received:

```
internal class DicomFileScope : IDisposable
{
    public DicomFileScope(DicomFile dicomFile) => File = dicomFile.Clone();
    public DicomFile File { get; }

    #region IDisposable Support

    private bool _isDisposed = false;

    protected virtual void Dispose(bool disposing)
    {
        if (!_isDisposed)
        {
            if (disposing)
            {
                File.Dataset?.Clear();
                File.FileMetaInfo?.Clear();
            }
            _isDisposed = true;
        }
    }

    public void Dispose() { Dispose(true); }

    #endregion
}
```

And for convenience, I've defined an extension method:

`public static DicomFileScope BeginScope(this DicomFile dicomFile) => new DicomFileScope(dicomFile);`

This can then be used around request processing:

```
/* ... */
using (var fileScope = dicomFile.BeginScope())
{
    var dicomCStoreRequest = new DicomCStoreRequest(fileScope.File);
    var dicomStatusTaskCompletionSource = new TaskCompletionSource<DicomStatus>();
    DicomCStoreRequest.ResponseDelegate responseReceived = (request, response) => 
        dicomStatusTaskCompletionSource.SetResult(response.Status);
    dicomCStoreRequest.OnResponseReceived += responseReceived;
    await _dicomClient.AddRequestAsync(dicomCStoreRequest);

    _ = _dicomClient.SendAsync(cancellationToken, DicomClientCancellationMode.ImmediatelyReleaseAssociation);

    var status = await dicomStatusTaskCompletionSource.Task;

    dicomCStoreRequest.OnResponseReceived -= responseReceived;
}
/* ... */
```

**Comment by amoerie** [comment]
You are right! I think the _sentRequests, at some point, were used for more than just its count, but the current situation seems to allow replacing that concurrent bag with a simple number indeed.

**Comment by amoerie** [comment]
This issue will be fixed for 4.0.6 and for 5.0.0
